src/dpl/src/AbacusLegalizer.cpp

```diff
@@ -0,0 +1,377 @@
+      inst->setLocation(coordinate.first, coordinate.second);
+    }
+  }
+
+  return cost;
+}
+
+void AbacusLegalizer::addCell(AbacusLegalizer::AbacusCluster* cluster,
+                              odb::dbInst* inst)
+{
+  auto instX = static_cast<double>(inst->getLocation().x());
+  auto instWidth = static_cast<double>(inst->getMaster()->getWidth());
+  double instE;
+  if (inst->getPlacementStatus().isFixed()) {
+    instE = std::numeric_limits<double>::max();
+  } else {
+    instE = 1;
+  }
+
+  cluster->instSet.push_back(inst);
+  cluster->e += instE;
+  cluster->q += instE * (instX - cluster->w);
+  cluster->w += instWidth;
+}
+
+void AbacusLegalizer::addCluster(AbacusLegalizer::AbacusCluster* predecessor,
+                                 AbacusLegalizer::AbacusCluster* cluster)
+{
+  predecessor->instSet.insert(predecessor->instSet.end(),
+                              cluster->instSet.begin(),
+                              cluster->instSet.end());
+  predecessor->e += cluster->e;
+  predecessor->q += cluster->q - cluster->e * predecessor->w;
+  predecessor->w += cluster->w;
+}
+
+void AbacusLegalizer::collapse(AbacusLegalizer::AbacusCluster& cluster,
+                               vector<AbacusCluster>& abacusClusters)
+{
+  auto xMin = (*targetBlock_->getRows().begin())->getBBox().xMin();
+  auto xMax = (*targetBlock_->getRows().begin())->getBBox().xMax();
+
+  cluster.x = cluster.q / cluster.e;  // place cluster c
+  // limit position between x min and x max - w_c(c)
+  if (cluster.x < xMin) {
+    cluster.x = xMin;
+  } else if (cluster.x > xMax - cluster.w) {
+    cluster.x = xMax - cluster.w;
+  }
+
+  if (abacusClusters.size() > 1) {
+    // if predecessor exists,
+    auto& predecessor = *(abacusClusters.end() - 2);
+    if (predecessor.x + predecessor.w > cluster.x) {
+      // merge cluster c to c'
+      addCluster(&predecessor, &cluster);
+      abacusClusters.erase(abacusClusters.end() - 1);  // remove cluster c
+      collapse(predecessor, abacusClusters);
+    }
+  }
+}
+
+void AbacusLegalizer::initRow()
+{
+  auto numRows = targetBlock_->getRows().size();
+  rows_.resize(numRows);
+  for (auto inst : targetBlock_->getInsts()) {
+    // for debugging
+    auto instName = inst->getName();
+
+    auto rowSample = (*targetBlock_->getRows().begin());
+    auto rowIdx = getRowIdx(inst);
+    auto rowHeight = rowSample->getBBox().dy();
+    auto yMin = rowSample->getBBox().yMin();
+    inst->setLocation(inst->getLocation().x(), rowIdx * rowHeight + yMin);
+    if (rowIdx >= 0 && rowIdx < numRows) {
+      if (inst->isFixed()) {
+        continue;
+      }
+      rows_.at(rowIdx).push_back(inst);
+      cellToRowMap_[inst] = &rows_.at(rowIdx);
+    }
+  }
+  for (int i = 0; i < rows_.size(); ++i) {
+    auto& row = rows_.at(i);
+    rowToRowIdxMap_[&row] = i;
+  }
+  // sort the instances in the row according to the x-coordinates for each row
+  for (auto& row : rows_) {
+    std::sort(
```

Comment:
It would be better to use stable_sort in case you have cells with equal x values for which this is not a proper ordering